common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java

```diff
@@ -203,16 +215,15 @@ private synchronized boolean shouldRetry(Throwable e) {
     boolean isIOException = e instanceof IOException
       || e.getCause() instanceof IOException;
     boolean isSaslTimeout = enableSaslRetries && e instanceof SaslTimeoutException;
-    if (!isSaslTimeout && saslTimeoutSeen) {
-      retryCount = 0;
-      saslTimeoutSeen = false;
+    // If this is a non SASL request failure, reduce earlier SASL failures from retryCount
+    // since some subsequent SASL attempt was successful
+    if (!isSaslTimeout && saslRetryCount > 0) {
+      retryCount -= saslRetryCount;
```

Comment:
Should we also verify that `retryCount >= saslRetryCount`? Just so that we don't ever end up in some funky state where we end up with negative values (I don't think this can ever happen, but would good to add the check!). :)

Reply:
I added a runtime exception.
But as you said, subset should never be bigger than the whole.

Reply:
looks good, thanks